[material-ui][ToggleButtonGroup] Support different elements under it

Methuselah96 opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Uses context for passing props from `ToggleButtonGroup` to `ToggleButton`s in the vein of https://github.com/mui/material-ui/pull/28645.

Fixes #36270
Fixes #12921

---

UI Regression test component image:

![Screenshot (19)](https://github.com/mui/material-ui/assets/20900032/f9df3654-e7a6-47fd-970b-ac3299eb9c30)

ZeeshanTamboli reviewed (commented):
After quickly reviewing the implementation, it seems satisfactory. However, could you include tests to confirm that the changes are functional? Specifically, test scenarios involving Toggle buttons wrapped with tooltips, ensuring that a toggle button is selected, and that styles are maintained when tooltips are wrapped around disabled buttons with `span`.

Also, I believe it closes both #12921 and #36270, so can you update the description to automatically close it when merged?

Methuselah96 commented:
I wasn't sure if #12921 was just about ToggleButtonGroup or if it's about replacing usages of `cloneElement` more broadly, do you still want me to mark it as closing that issue?

Methuselah96 commented:
I added the changes similar to #38520 and #38989 to fix the styling for varying children and added some tests as well.

The current issue is that sometimes the negative margin CSS gets overridden, by either `ToggleButton` or by custom styles. Any ideas on how to make sure that the negative margin takes precedence, especially over the margin set in `ToggleButton`?

Methuselah96 commented:
Alright, I overcame that issue by using a more specific selector to make sure the toggle button border styling is preserved even when the button is disabled.

Everything should be ready to go, let me know if there's anything else you need!

ZeeshanTamboli commented on docs/data/material/components/toggle-button/CustomizedDividers.tsx:
```diff
@@ -18,15 +18,18 @@ const StyledToggleButtonGroup = styled(ToggleButtonGroup)(({ theme }) => ({
-      borderRadius: theme.shape.borderRadius,
-    },
+  },
+  '& .MuiToggleButtonGroup-middleButton': {
+    marginLeft: -1,
+    borderLeft: '1px solid transparent',
+  },
+  '& .MuiToggleButtonGroup-lastButton': {
+    marginLeft: -1,
+    borderLeft: '1px solid transparent',
```
We can combine this:
```suggestion
  '& .MuiToggleButtonGroup-middleButton,& .MuiToggleButtonGroup-lastButton': {
    marginLeft: -1,
    borderLeft: '1px solid transparent',
  },
```
Also, now, is this considered a breaking change for developers using those pseudo-classes (:first-of-type, :last-of-type) for custom styling?

Methuselah96 replied:
Yes, I believe it is a slight breaking change, because the pseudo-classes were more specific than the new classes, and therefore took higher precedence. I think https://github.com/mui/material-ui/pull/38520 also would've also been a slight breaking change in that sense?

ZeeshanTamboli replied:
I believe merging this PR is worthwhile for resolving the associated issue (considering the upvotes in #12921), even if it involves a minor risk of breaking changes. Given our successful inclusion of #38520 in a previous patch release without any issues, I think we can do the same with this one.

My only concern is that developers who have copied this demo directly into their projects may be affected by the changes in this PR.

Let's keep this conversation thread open even after the PR gets merged.

ZeeshanTamboli commented on packages/mui-material/src/ToggleButtonGroup/ToggleButtonGroupContext.ts:
```diff
@@ -0,0 +1,23 @@
+interface IToggleButtonGroupContext {
+  className?: string;
+  onChange?: ToggleButtonGroupProps['onChange'];
+  value?: any;
```
```suggestion
  value?: ToggleButtonGroupProps['value'];
```

ZeeshanTamboli commented on packages/mui-material/src/ToggleButtonGroup/ToggleButtonGroupContext.ts:
```diff
@@ -0,0 +1,23 @@
+  onChange?: ToggleButtonGroupProps['onChange'];
+  value?: any;
+  size?: ToggleButtonGroupProps['size'];
+  fullWidth?: boolean;
```
```suggestion
  fullWidth?: ToggleButtonGroupProps['fullWidth'];
```

ZeeshanTamboli commented on packages/mui-material/src/ToggleButtonGroup/ToggleButtonGroupContext.ts:
```diff
@@ -0,0 +1,23 @@
+  size?: ToggleButtonGroupProps['size'];
+  fullWidth?: boolean;
+  color?: ToggleButtonGroupProps['color'];
+  disabled?: boolean;
```
```suggestion
  disabled?: ToggleButtonGroupProps['disabled'];
```

ZeeshanTamboli commented on packages/mui-material/src/ToggleButtonGroup/toggleButtonGroupClasses.ts:
```diff
@@ -16,6 +16,12 @@ export interface ToggleButtonGroupClasses {
   groupedVertical: string;
   /** Styles applied to the root element if `fullWidth={true}`. */
   fullWidth: string;
+  /** Styles applied to the first button in the button group. */
+  firstButton: string;
+  /** Styles applied to the last button in the button group. */
+  lastButton: string;
+  /** Styles applied to buttons in the middle of the button group. */
+  middleButton: string;
```
To be more specific:
```suggestion
  /** Styles applied to the first button in the toggle button group. */
  firstButton: string;
  /** Styles applied to the last button in the toggle button group. */
  lastButton: string;
  /** Styles applied to buttons in the middle of the toggle button group. */
  middleButton: string;
```

ZeeshanTamboli reviewed (commented):
> I wasn't sure if https://github.com/mui/material-ui/issues/12921 was just about ToggleButtonGroup or if it's about replacing usages of cloneElement more broadly, do you still want me to mark this PR as closing that issue?

After reviewing the issue thread, I can confirm it is related to the toggle button group. We can mark the PR as resolving that issue.

Methuselah96 commented:
PR review comments addressed.

ZeeshanTamboli reviewed (approved):
Looks good!
